prune/prune.go

```diff
@@ -17,181 +17,197 @@ package prune
+	// true = Just check, don't prune
+	// false (default) = Prune
+	DryRun bool
+
+	// Strategy is the function used to determine a list of resources that are pruneable
+	Strategy StrategyFunc
 
-// StrategyConfig holds settings unique to each pruning mode
-type StrategyConfig struct {
-	Mode            Strategy
-	MaxAgeSetting   string
-	MaxCountSetting int
-	CustomSettings  map[string]interface{}
+	// Labels is a map of the labels to use for label matching when looking for resources
+	Labels map[string]string
+
+	// Namespace is the namespace to use when looking for resources
+	Namespace string
+
+	// Logger is the logger to use when running pruning functionality
+	Logger logr.Logger
 }
 
-// StrategyFunc function allows a means to specify
-// custom prune strategies
-type StrategyFunc func(ctx context.Context, cfg Config, resources []ResourceInfo) ([]ResourceInfo, error)
-
-// PreDelete function is called before a resource is pruned
-type PreDelete func(ctx context.Context, cfg Config, something ResourceInfo) error
-
-// Config defines a pruning configuration and ultimately
-// determines what will get pruned
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// Question: Where should this be used? Why should this be used over the Go default error?
+// ErrUnpruneable indicates that it is not allowed to prune a specific object.
+type ErrUnpruneable struct {
+	Obj    *client.Object
+	Reason string
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
+
+// IsPruneableFunc is a function that checks a the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `ErrUnpruneable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPruneableFunc func(obj client.Object) error
+
+// PrunerOption configures the pruner.
+type PrunerOption func(p *Pruner)
+
+// Error returns a string reprenstation of an `ErrUnpruneable` error.
```

Comment:
```suggestion
// Error returns a string representation of an `ErrUnpruneable` error.
```

Reply:
fixed in https://github.com/operator-framework/operator-lib/pull/105/commits/1fca33a0903e9f38f9f39390c315c8618bbc2414